File: scala-stm/src/main/scala/io/github/gvolpe/transfer2.scala
package io.github.gvolpe

import scalaz.zio._
import scalaz.zio.clock.Clock
import scalaz.zio.duration._
import scalaz.zio.stm._
import transfer._

object transfer2 {

  def withdraw(account: Account, amount: Int): STM[Nothing, Unit] =
    account.ref.get.flatMap { balance =>
      STM.check(amount <= 0 || amount <= balance) *>
        account.ref.set(balance - amount)
    }

  def delayDeposit(account: Account, amount: Int): ZIO[Clock, Nothing, Unit] =
    for {
      _ <- putStrLn("Getting ready to deposit money...hunting through pockets...")
      _ <- ZIO.sleep(3.seconds)
      _ <- putStrLn("OK! Depositing now!")
      _ <- STM.atomically(account.ref.update(_ + amount))
    } yield ()

  def main: ZIO[Clock, Nothing, Unit] =
    for {
      from <- STM.atomically(TRef.make(100)).map(Account)
      _    <- delayDeposit(from, 1).fork.void
      _    <- putStrLn("Trying to withdraw money...")
      _    <- STM.atomically(withdraw(from, 101))
      _    <- putStrLn("Successful withdrawal!")
    } yield ()

}

File: haskell-stm/src/Transfer2.hs
module Transfer2
  ( main2
  )
where

import           Control.Monad                  ( when )
import           Control.Concurrent.STM
import           Control.Concurrent             ( forkIO
                                                , threadDelay
                                                )

{- 2nd example showing a late deposit and a withdrawal waiting for the amount to be present -}

type Account = TVar Int

withdraw :: Account -> Int -> STM ()
withdraw acc amount = do
  bal <- readTVar acc
  check (amount <= 0 || amount <= bal)
  writeTVar acc (bal - amount)

delayDeposit :: Account -> Int -> IO ()
delayDeposit acc amount = do
  putStrLn "Getting ready to deposit money...hunting through pockets..."
  threadDelay 3000000
  putStrLn "OK! Depositing now!"
  atomically
    (do
      bal <- readTVar acc
      writeTVar acc (bal + amount)
    )

main2 = do
  acc <- atomically (newTVar 100)
  forkIO (delayDeposit acc 1)
  putStrLn "Trying to withdraw money..."
  atomically (withdraw acc 101)
  putStrLn "Successful withdrawal!"

